src/common/ask-boolean/ask-boolean-question-channel.injectable.ts

```diff
@@ -0,0 +1,22 @@
+/**
+ * Copyright (c) OpenLens Authors. All rights reserved.
+ * Licensed under MIT License. See LICENSE in root directory for more information.
+ */
+import { getInjectable } from "@ogre-tools/injectable";
+import type { Channel } from "../channel/channel-injection-token";
+import { channelInjectionToken } from "../channel/channel-injection-token";
+
+export interface AskBooleanQuestionParameters { id: string; title: string; question: string }
+export type AskBooleanQuestionChannel = Channel<AskBooleanQuestionParameters>;
+
+const askBooleanQuestionChannelInjectable = getInjectable({
+  id: "ask-boolean-question-channel",
+
+  instantiate: (): AskBooleanQuestionChannel => ({
+    id: "ask-boolean-question",
+  }),
+
+  injectionToken: channelInjectionToken,
+});
```

Comment:
Main doesn't take any responsibility from how it is asked from user. In main, you will just specify that `askBoolean` and you don't care how it's done. As long as it returns promise that resolves with the boolean value. It doesn't even care who answers, it might be one window or the other. Or even the main it self.

Reply:
> really breaks the separation of concern and decoupling that we are working so hard to remove

There is nothing like that here. What do you mean?

Remember that this `askBooleanQuestionChannel` is internal technical detail and not that important. Important part is `askBoolean` and the contract that it represents.

Reply:
Main shouldn't even know that `askBoolean` is a possible contract.

Reply:
And why would that be? In perfect world we are able to develop new features without directory jumping between `main` and `renderer`. Long term this is achievable by making things comply with Open Closed Principle and then creating hard boundaries. This allows us to move to direction where features (behaviours) are present in e.g. directory structures and they can be removed simply by removing everything from the directory. (or even separate NPM package).

That being said, `main`s only real purpose is to handle stuff that shouldn't be handled in window (for whatever reason) for the feature. Which means that it shouldn't be overly generic and try to serve all features.

Reply:
You need to provide me with what you mean by OCP because all I can read about it has nothing todo with what you are using it as.

Furthermore, you haven't made a good argument as to why main should know about the rendering of values, of even why it should make the decisions on how it is displayed, when, or in how many locations.

Reply:
> You need to provide me with what you mean by OCP because all I can read about it has nothing todo with what you are using it as.

Open Closed Principle describes that you should be able add functionality without modifying the existing code. 

-----

#### Example 1

I have requirement to add new item in a tray (context irrelevant; add anything, anywhere.) Minimal PR contains only additions, no modifications and no deletions. 


#### Example 2

I have requirement to remove some item from a tray (context irrelevant; remove anything, anywhere.) Minimal PR contains only deletions, no modifications and no additions.

-----

Making code to comply with it has a lot of benefits. One that I was describing in previous message is step towards achieving something called Screaming Architecture from Clean Architecture.

Reply:
Sure, main should run code until it has been told to. I am not trying to argue with that.

My concern with this code is the explicitness of main asking questions and dictating that. You could still have main not do any more work if instead main just let renderer know that there is an update and have a well defined place for renderer to let main know that it should continue. That way we have more decoupling as well.

Reply:
Yes I know that the behaviours are good and help use to more fearless refactorings.

About 3, just because they don't have to register new listeners doesn't mean that it is better. Now `main` has to start caring about UX.

We can create a similar level of abstraction without this sort of thing, something like the following maybe:

```ts
autoUpdate.once("user-wants-to-update", () => doUpdate());
emitUpdateReadyForInstallation();
```

> There is no stake in how that happens, as something else decides that. Therefore, there is no coupling, and we are free to define or re-define how the contract is fulfilled.

No arguments there, my concern is that `main` shouldn't care about UX that is not its responsibility.

> We've seen something like askBoolean used many times before with only good things to say about it. We have good confidence that askBoolean will not cause trouble. And even if it does, we will still have the behaviours to enable confident and effective refactoring.

Have you done it from a website's backend to the frontend?

Reply:
Resolved this conversation privately by making sure that everyone understands that we are not looking for solutions that are written in stone. We are looking for solutions that can later be improved easily when the requirement comes. 

That being said, for now, it's good enough since behaviour is green. Future requirements will re-shape the implementation if needed.